src/sources/kubernetes.rs

```diff
@@ -0,0 +1,318 @@
+    config.data_dir = Some(globals.resolve_and_make_data_subdir(None, kube_name)?);
+
+    let (file_send, file_recv) = mpsc::channel(1000);
+    let file_source = config
+        .build("file_source", globals, file_send)
+        .map_err(|e| format!("Failed in creating file source with error: {:?}", e))?;
+
+    Ok((file_recv, file_source))
+}
+
+fn parse_message() -> Result<impl FnMut(Event) -> Option<Event> + Send, String> {
+    // Transforms
+    let mut transform_json_message = transform_json_message();
+    let mut transform_cri_message = transform_cri_message()?;
+
+    // Kubernetes source
+    let atom_pod_uid = Atom::from("pod_uid");
+    let mut pod_data = HashMap::<Bytes, PodData>::new();
+    Ok(move |event: Event| {
+        let log = event.as_log();
+        if let Some(ValueKind::Bytes(pod_uid)) = log.get(&atom_pod_uid) {
+            // Fetch pod data
+            let data = pod_data.entry(pod_uid.clone()).or_default();
+
+            // Get/determine log format
+            let log_format = if let Some(log_format) = data.log_format {
+                log_format
+            } else {
+                // Detect log format
+                let log_format = if transform_json_message(event.clone()).is_some() {
+                    LogFormat::Json
+                } else if transform_cri_message.transform(event.clone()).is_some() {
+                    LogFormat::CRI
+                } else {
+                    error!(message = "Unknown message format");
+                    // Return untouched message so that user has some options
+                    // in how to deal with it.
+                    LogFormat::Unknown
+                };
+                // Save log format
+                data.log_format = Some(log_format);
+                log_format
+            };
+
+            // Parse message
+            match log_format {
+                LogFormat::Json => transform_json_message(event),
+                LogFormat::CRI => transform_cri_message.transform(event),
+                LogFormat::Unknown => Some(event),
+            }
+        } else {
+            None
+        }
+    })
+}
+
+fn transform_json_message() -> impl FnMut(Event) -> Option<Event> + Send {
+    let mut config = JsonParserConfig::default();
+
+    // Drop so that it's possible to detect if message is in json format
+    config.drop_invalid = true;
+
+    // Drop field
+
+    let mut json_parser: JsonParser = config.into();
+
+    let atom_time = Atom::from("time");
+    let atom_log = Atom::from("log");
+    move |event| {
+        let mut event = json_parser.transform(event)?;
+
+        // Rename fields
+        let log = event.as_mut_log();
+
+        // time -> timestamp
+        if let Some(ValueKind::Bytes(timestamp_bytes)) = log.remove(&atom_time) {
+            match DateTime::parse_from_rfc3339(
+                String::from_utf8_lossy(timestamp_bytes.as_ref()).as_ref(),
+            ) {
+                Ok(timestamp) => log.insert_explicit(
```

Comment:
I believe this should be implicit because we are adding it?

Reply:
We are adding it, but the timestamp is taken by the container runtime, so this is a question of is that explicit or implicit? I had an opinion on this, but I forgot it. In the docker source it's explicit, so we should also change that if we decide the opposite.

I have taken a look at the other sources that add timestamps gotten from other side, and `journald source` is the only one of them, and it is adding timestamp as explicit.